cstrml/benefits/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+                    fee_reduction.total_fee_reduction_count = Self::calculate_total_count_reduction(&fee_reduction.funds);
+                }
+            );
+            <OverallBenefits<T>>::mutate(|benefits| { benefits.total_funds = benefits.total_funds.saturating_sub(value.clone());});
+
+            // 3. Emit success
+            Self::deposit_event(RawEvent::CutBenefitFundsSuccess(who.clone(), value));
+
+            Ok(())
+        }
+    }
+}
+
+
+impl<T: Config> Module<T> {
+    /// The return value is the used fee quota in the last era
+    pub fn update_era_benefit_inner(next_era: EraIndex, total_benefits: BalanceOf<T>) -> BalanceOf<T> {
+        // Fetch overall benefits information
+        let mut overall_benefits = Self::overall_benefits();
+        // Store the used fee reduction in the last era
+        let used_benefits = overall_benefits.used_benefits;
+        // Start the next era and set active era to it
+        overall_benefits.active_era = next_era;
+        // Set the new total benefits for the next era
+        overall_benefits.total_benefits = total_benefits;
+        // Reset used benefits to zero
+        overall_benefits.used_benefits = Zero::zero();
+        <OverallBenefits<T>>::put(overall_benefits);
+        // Return the used benefits in the last era
+        used_benefits
+    }
+
+    pub fn maybe_free_count_inner(who: &T::AccountId) -> bool {
+        let overall_benefits = Self::overall_benefits();
+        let mut fee_reduction = Self::fee_reduction_ledger(who);
+        Self::try_refresh_fee_reduction(&overall_benefits, &mut fee_reduction);
+        // won't update reduction detail if it has no staking
+        // to save db writing time
+        if fee_reduction.used_fee_reduction_count < fee_reduction.total_fee_reduction_count {
+            fee_reduction.used_fee_reduction_count += 1;
+            <FeeReductionLedger<T>>::insert(&who, fee_reduction);
+            return true;
+        }
+        return false;
+    }
+
+    pub fn maybe_reduce_fee_inner(who: &T::AccountId, fee: BalanceOf<T>, reasons: WithdrawReasons) -> Result<NegativeImbalanceOf<T>, DispatchError> {
+        let mut overall_benefits = Self::overall_benefits();
+        let mut fee_reduction = Self::fee_reduction_ledger(who);
+        // Refresh the reduction
+        Self::try_refresh_fee_reduction(&overall_benefits, &mut fee_reduction);
+        // Calculate the own reduction limit
+        let fee_reduction_benefits_quota = Self::calculate_total_benefits(fee_reduction.funds,
+                                                                overall_benefits.total_funds,
+                                                                overall_benefits.total_benefits);
+        // Try to free fee reduction
+        // Check the person has his own fee reduction quota and the total benefits
+        let benefit_costs = T::BenefitMarketCostRatio::get() * fee;
```

Comment:
Super unclear with this naming, If Cost is `5%`, it should named `charged_fee` not `benefit_cost`
The current `fee_reduction_benefits` should named `fee_reduction_benefit_cost`

Reply:
And if I were you, I'll write set this `BenefitCost` as `95%`, and only calculate the charged_fee when the `if` is true